keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+part of this KEP.
+
+Two new API object types get added in a new API group:
+- ResourceClass, not namespaced, with the name of a resource driver and
+  parameters for the resource driver that will be used for all ResourceClaims
+  that reference this ResourceClass. These parameters might describe the kind
+  of resource that are allocated when using the ResourceClass or control
+  cluster-specific options. Because only cluster administrators are allowed to
+  create ResourceClasses and they will be passed separately to the resource
+  driver, they can be used for options that normal users should not be allowed
+  to control.
+
+  For example, a single resource driver might manage different kinds of
+  FPGAs. For each kind, one ResourceClass could define the kind and provide paths
+  to kind-specific tools or URLs for additional resources. Only a handful of
+  ResourceClasses are expected per resource driver.
+
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual metadata, a spec and a status. The
+  spec identifies the driver that handles the resource via a class
+  name.
+
+A third one with information about the driver ("ResourceDriver", similar to the
+"CSIDriver" from storage) could get added in the future. At the moment it is
+not needed yet.
+
+The ResourceClaim spec is immutable. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost, which in the past was not ruled
+out. For example, status could have been stored in a separate etcd instance
+with lower reliability. To recover after a loss, status was meant to be recoverable.
+A [recent KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac)
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+[runtime.RawExtension](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension) (see [user stories](#user-stories) for examples).
+Validation can be handled by drivers through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the driver.
+
+The reasons for choosing embedded RawExtension fields fields instead of
+references to a custom resource (CR) that is defined by a driver's custom
+resource definition (CRD) are:
+
+- Immutability can be enforced: a separate object could always get
+  deleted and recreated with new content.
+- More user-friendly because there is no need to manage an additional
+  object.
+- Less risky resource driver deployment and updates: CRs get deleted when
+  removing their CRD, which would render existing workloads unusable until
+  users redeploy them.
```

Comment:
Drawbacks of this proposal:
- Validation is limited to those APIs for which the resource driver installs validating webhooks. Given that parameters may be embedded inside a PodSpec which itself is embedded in other, potentially unknown, APIs, this is never going to be complete.
- Validation has to be done via a webhook.

A reference to a CR would be just the object name. For ResourceClass, that object is cluster-scoped. For ResourceClaim, it is in the same namespace as the ResourceClaim. Do we need a kind and version or can resource drivers figure that out themselves? I think they can, as they most likely will just have one kind+version.

Immutability could be enforced by the resource driver: at the point where it allocates the resource, it does so with the parameters at that time. It must then ignore the CR and be able to use and free the resource without that CR, just as it must be able to free without the ResourceClass.

A resource driver might also decide to support resizing by reconciling future parameter changes with an allocated resource. How it does that would be outside of this KEP. This KEP does not require that the status is immutable after allocation, so the driver might even be able to update some of its fields (availability, number of users). This must not affect running Pods, though, so it cannot take away resources.

I still think embedded parameters are easier to use, except for the validation problem. Perhaps embedding a CR would be ideal, but I don't think that this is an option at the moment.

Deleting CRs is a problem, but it's a general Kubernetes problem. Admins will have to be careful...

Reply:
Let's use [`v1.ObjectReference`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#objectreference-v1-core) instead of `RawExtension`.

It is a large struct with some fields that are probably not needed (for example, UID). In the updated API description below I have tried to explain which fields are relevant and what additional constraints apply (mostly around the namespace).

We should encourage resource driver authors and users to prefer "kind" over "resource" (see https://github.com/kubernetes/community/pull/5973), but the API will support both.